### PR TITLE
Stardew Valley - Added missing logic rules for dating and marriage

### PR DESCRIPTION
**Pull request by agilbert1412**
## What is this fixing or adding?
When playing on friendsanity marriage, for some values of heart size, "milestone" hearts can end up being skipped. For example, with heart size 3, the hearts are 3-6-9-12-14, which skips 8 and 10, which are the milestones for dating and marriage.

In order to "earn" heart 9, you need to have 8 hearts and give a bouquet to the NPC to date them. Under the current logic, only heart 6 would be required for heart 9 to be in logic, as the design was "previous heart only". This is a problem, because at 6 hearts, you cannot date the villager, and therefore you cannot "earn" hearts past the 8th one.

This fixes the problem. Any heart earned past 8 requires dating, so 8 minimum received, and any past 10 requires marriage, so 10 minimum received. Taking the same example, in order to "earn" heart 9, you would now need to receive heart 9, in order to give a bouquet.

## How was this tested?
Created a new unit test that finds that specific bug before fixing it, then fixed it and ran all of the tests

### CODE REVIEW

File: worlds/stardew_valley/logic.py [hunk 1]
```diff
@@ -1081,8 +1081,7 @@ def has_relationship(self, npc: str, hearts: int = 1) -> StardewRule:
             return self.can_earn_relationship(npc, hearts)
         if self.options[options.Friendsanity] == options.Friendsanity.option_starting_npcs and not villager.available:
             return self.can_earn_relationship(npc, hearts)
-        if self.options[
-            options.Friendsanity] != options.Friendsanity.option_all_with_marriage and villager.bachelor and hearts > 8:
+        if self.options[options.Friendsanity] != options.Friendsanity.option_all_with_marriage and villager.bachelor and hearts > 8:
```
Comment (black-sliver) [inline worlds/stardew_valley/logic.py:1084]: for the future: it is preferred to stay within 120 chars per line. pycharm normally does good automated breaking, but sometimes doing a manual break with `\` looks better, like
```python
        if self.options[options.Friendsanity] != \
                options.Friendsanity.option_all_with_marriage and villager.bachelor and hearts > 8:
            ...
```
for example

Comment (agilbert1412) [reply worlds/stardew_valley/logic.py]: Personally, my Pycharm is configured for 160, not 120, as that is my preference, but I fixed it anyway, as I do agree this one line was difficult to read, and the length was a factor in that

**Comment by black-sliver** [comment]
I'll look at it (test, merge) tonight.

160 sounds like a lot.  The 120 is actually already more than is easy to read, however with all the leading white spaces and possibly trailing characters, 120 is still okay-ish most of the time and does not introduce as many forced line breaks as 80 (that also impact readability). This is not just how much you can fit onto your screen layout, but also how far the eyes have to jump when browsing through code.

**Review by black-sliver — APPROVED** [review]
lgtm

**Pull request merged**